cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -1119,41 +1119,67 @@ def _get_raw_x_loc(self) -> str:
         else:
             return "X"
 
-    def _is_raw(self, force: bool = False) -> bool:
+    def _has_valid_raw(self, force: bool = False) -> bool:
         """
         Checks if the non-zero values for the raw matrix (adata.X or adata.raw.X)
-        are integers. Returns False if at least one value is not an integer,
+        are positive integers stored as numpy.float32. Also validates that every row contains at least one non-zero
+        value.
 
-        True otherwise.
+        Returns False if at least one value / row does not meet requirements. True otherwise.
 
         Since this process is memory intensive, it will return a cache value if this function has been called before.
         If calculation needs to be repeated use `force = True`
 
         :rtype bool
-        :return False if at least one value is not an integer, True otherwise
         """
         if force:
             self._raw_layer_exists = None
 
         if self._raw_layer_exists is None:
             # Get potential raw_X
-            raw_loc = self._get_raw_x_loc()
-            x = self.adata.raw.X if raw_loc == "raw.X" else self.adata.X
+            x = self.adata.raw.X if self._get_raw_x_loc() == "raw.X" else self.adata.X
+            if x.dtype != np.float32:
+                self._raw_layer_exists = False
+                self.errors.append("Raw matrix values must have type numpy.float32.")
+                return self._raw_layer_exists
 
             matrix_format = get_matrix_format(self.adata, x)
             assert matrix_format != "unknown"
             self._raw_layer_exists = True
+            has_row_of_zeros = False
+            has_invalid_nonzero_value = False
+            is_sparse_matrix = matrix_format in SPARSE_MATRIX_TYPES
             for matrix_chunk, _, _ in self._chunk_matrix(x):
```

Comment:
@nayib-jose-gloria I think we have a problem here cc @Bento007 

This is returning a _chunk_ of the matrix, therefore a possible subset of all values. But the logic in the loop appears to be treating it as if it comprises _all_ values. We're comparing the n_obs of the entire dataset to the number of rows represented in the chunk, and if they aren't equal we're asserting that a row of zeros must exist. This seems like a mistake to me.

Reply:
great catch--I believe this wouldn't have been caught in QA because the chunk size is 10,000 rows and the test datasets are all smaller than that. 

I can make the fix

Reply:
>this wouldn't have been caught in QA because the chunk size is 10,000 rows and the test datasets are all smaller than that

Yeah I had suspected that would be the case.

Btw do you know much about the `_chunk_matrix` function? Does using the `[]` indexing notation to read off of a matrix not actually load the data into mem until you make that call? @bkmartinjr 's comments in the docstring about "read fast-path" are something I'm not familiar with.

Reply:
@nayib-jose-gloria you might consider pulling in the logic loop changes from my original feature branch. Or did you read over it and decide it needed to be overhauled..? let me know

Reply:
I did this somewhere else, I think for seurat convertability, but you should make the default chunk size in `_chunk_matrix` a variable that you can patch in the unit tests. That way you can test compatibility with datasets whose size is greater than the chunk size while still using the same test dataset(s).